content/docs/rfcs/13/README.md

```diff
@@ -59,139 +59,260 @@ The following are not considered as part of the adversarial model:
 In specific, the communication channels are assumed to be secure.
 
 # Wire Specification
-Peers communicate with each other using a request / response API. 
-The messages sent are Protobuf RPC messages which are implemented using [protocol buffers v3](https://developers.google.com/protocol-buffers/).
-The following are the specifications of the Protobuf messages.
 
 ## Payloads
 
 ```protobuf
 syntax = "proto3";
 
-message Index {
-  bytes digest = 1;
-  sint64 receiverTime = 2;
-  sint64 senderTime = 3;
-  string pubsubTopic = 4;
-}
-
-message PagingInfo {
-  uint64 pageSize = 1;
-  Index cursor = 2;
-  enum Direction {
-    BACKWARD = 0;
-    FORWARD = 1;
-  }
-  Direction direction = 3;
-}
+// Protocol identifier: /vac/waku/store/3.0.0
+package waku.store.v3;
 
-message ContentFilter {
-  string contentTopic = 1;
-}
+import "waku/message/v1/message.proto";
 
-message HistoryQuery {
-  // the first field is reserved for future use
-  string pubsubtopic = 2;
-  repeated ContentFilter contentFilters = 3;
-  PagingInfo pagingInfo = 4;
+message WakuMessageKeyValue {
+  optional bytes message_hash = 1; // Globally unique key for a Waku Message
+  optional waku.message.v1.WakuMessage message = 2; // Full message content as value
 }
 
-message HistoryResponse {
-  // the first field is reserved for future use
-  repeated WakuMessage messages = 2;
-  PagingInfo pagingInfo = 3;
-  enum Error {
-    NONE = 0;
-    INVALID_CURSOR = 1;
-  }
-  Error error = 4;
+message StoreRequest {
+  string request_id = 1;
+  bool return_values = 2; // Response should include full message content
+  
+  // Filter criteria for content-filtered queries
+  optional string pubsub_topic = 10;
+  repeated string content_topics = 11;
+  optional int64 time_start = 12;
+  optional int64 time_end = 13;
+
+  // List of key criteria for lookup queries
+  repeated bytes message_hashes = 20 // Message hashes (keys) to lookup
+  
+  // Pagination info. 50 Reserved
+  optional bytes pagination_cursor = 51; // Message hash (key) from where to start query (exclusive)
+  bool pagination_forward = 52;
+  optional uint64 pagination_limit = 53;
 }
 
-message HistoryRPC {
+message StoreResponse {
   string request_id = 1;
-  HistoryQuery query = 2;
-  HistoryResponse response = 3;
+
+  optional uint32 status_code = 10;
+  optional string status_desc = 11;
+
+  repeated WakuMessageKeyValue messages = 20;
```

Comment:
Maybe a `map` ?

Reply:
After coding the nwaku impl. I think that it would be harder to do.

Reply:
Indeed. Especially if only one or the other field is present.